### ISSUE TITLE
accessibility labels on ios 10(xcode 8 beta 6) are not recognized

### ISSUE THREAD

**Issue by Tgillon** [kif-framework/KIF#897]
it seems that kif tests which run on ios 10 fail since accessibility labels are not recognized any more.
if a ui view has a title it will be recognized as his accessbilityLabel.
for example :
view.setTitle="keypad"
view.accessibilityLabel="dialer"

if i use the inspector I'll get that the accessiblity label is "keypad" and not "dialer"

**Comment by Tgillon** [comment]
Hi,


I just checked.
Accessibility Identifier works as in iOS 9.
Thanks,
Tali

**Comment by hanneskaeufler** [comment]
I'm tripping on this as well when upgrading to ios10 and KIF 3.5.2.
I have some buttons in a `UIToolbar` which is added as an `inputAccessoryView` being shown with the keyboard.

The button has a

```swift
        doneButton.accessibilityLabel = "doneButton"
```

This is how it shows up in the inspector

<img width="375" alt="bildschirmfoto 2016-09-18 um 21 12 38" src="https://cloud.githubusercontent.com/assets/68024/18618337/ffaf38aa-7de4-11e6-925d-8018fe331893.png">

All of my tests for this view are timing out, not finding the buttons.

**Comment by mikelupo** [comment]
Based off a previous conversation I think you can use accessibility identifiers as a work around.

@phatmann any insight on this problem?

**Comment by phatmann** [comment]
Since this is not a problem with KIF per se, I am going to close this issue. If anyone wants to create a PR with a workaround in KIF, I would be happy to look at it.

**Comment by kunalbalani** [comment]
Found a workaround. Accessibility inspector is moved to Xcode from simulator. If you enable that labels just start working fine. Just open Xcode -> Open Developer Tools -> Accessibility Inspector and run the test suite again.

**Comment by phatmann** [comment]
Interesting! Does that mean the KIF should not try to enable the Accessibility Inspector in the simulator in Xcode 8?

**Comment by mikelupo** [comment]
Wondering out loud... Would not enabling the AI have any affect on an actual device?

**Comment by phatmann** [comment]
I don’t think our AI-enabling hack works on devices.

Tony

**Comment by ecaselles** [comment]
We are having the same issues here (ping @emmaviolet) and have come to similar conclusions to the ones by @kunalbalani. 

Most of our tests fail when running on Xcode 8.1 and any iOS 10.1 simulators. Then, if we open the new Accessibility Inspector, attach it to the Simulator process and run the app (not even required to run the test). Then, we run the tests and they all start passing. After doing this once, running the tests again will work and tests will pass, even if the AI is closed. This workaround is "acceptable" for running the tests locally, but we have got no idea on how to make it work for our cloud CI. Any suggestions?

Just wondering, @phatmann, what is exactly KIF trying to do with the AI?

**Comment by kunalbalani** [comment]
Just swizzle AI with AL, it's test anyways. Do this and you are green.

```
+ (void)load
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        Class classView = [UIView class];
        Class classBarItem = [UIBarItem class];

        SEL originalSelector = @selector(accessibilityLabel);
        SEL swizzledSelector = @selector(accessibilityIdentifier);
        
        Method originalMethod = class_getInstanceMethod(classView, originalSelector);
        Method swizzledMethod = class_getInstanceMethod(classView, swizzledSelector);
        method_exchangeImplementations(originalMethod, swizzledMethod);
        
        originalMethod = class_getInstanceMethod(classBarItem, originalSelector);
        swizzledMethod = class_getInstanceMethod(classBarItem, swizzledSelector);
        method_exchangeImplementations(originalMethod, swizzledMethod);
        
        
        originalSelector = @selector(setAccessibilityLabel:);
        swizzledSelector = @selector(setAccessibilityIdentifier:);
        
        originalMethod = class_getInstanceMethod(classView, originalSelector);
        swizzledMethod = class_getInstanceMethod(classView, swizzledSelector);
        method_exchangeImplementations(originalMethod, swizzledMethod);
        
        originalMethod = class_getInstanceMethod(classBarItem, originalSelector);
        swizzledMethod = class_getInstanceMethod(classBarItem, swizzledSelector);
        method_exchangeImplementations(originalMethod, swizzledMethod);

    });
}

```

**Comment by glentregoning** [comment]
I would argue we should reopen this issue as this change in the way iOS provides accessibility information fundamentally breaks using accessibility labels with KIF on iOS 10. That is unless the go forward plan is to drop support for labels, and move to identifying by Id going forward?

**Comment by phatmann** [comment]
I understand that iOS's latest changes are making accessibility labels hard to use. However, making accessibility identifiers the only control identification mechanism in KIF would be a major breaking change. KIF allows users to use identifiers or labels, so why force people to use identifiers if labels are working for them? The real problem, I believe, is that KIF's mechanism for opening the accessibility inspector is apparently failing to work in Xcode 8.

I have re-opened this issue for more discussion per @glentregoning's request.

**Comment by justinseanmartin** [comment]
I concur that `accessibilityLabel` must continue to work, as this is the typical way of enabling accessibility within an application. Swizzling accessibilityLabel to pull the value from accessibilityIdentifier is unnecessary. If you're doing that, you might as well just change all your tests to pull the accessibilityIdentifier field instead. However, that defeats a large benefit of KIF, which is to verify the accessibility of your application.

It seems to work fine on Xcode 8.1 for me in the simulator. There have been a bunch of replies and it is unclear to me as to if everyone that is having issues is only on physical devices. The last few replies sound like simulator issues as well.

For people hitting issues in the simulator, is the code hitting the line in `KIFAccessibilityEnabler.m`:

```
loadDylibForSimulator(@"/usr/lib/libAccessibility.dylib");
```

Things seem to work fine for us without the AI enabled in Xcode 8. We have a slight fork at the moment internally, but I don't think it should be functionally any different than what is currently in GitHub master.

**Comment by glentregoning** [comment]
Thanks @justinseanmartin and @phatmann, upon further investigation we have resolved this issue by upgrading to the latest KIF and ensuring KIFAccessibilityEnabler is invoked as you suggested. Given this, from my perspective I think it would be ok to close this issue now.

**Comment by ecaselles** [comment]
At the end, after a bit of investigation yesterday, we got to the same conclusion explained by @justinseanmartin: the problem was that our tests were never hitting the fix introduced to enable the accessibility inspector in Xcode 8 (see [#877](https://github.com/kif-framework/KIF/pull/877) and [#883](https://github.com/kif-framework/KIF/pull/883)).

This was due to what @phatmann had already raised in #890: since we were using [KIF-Kiwi](https://github.com/indiegogo/KIF-Kiwi), our specs do not inherit directly from `KIFTestCase`, therefore the line to [enable accessibility on set up](https://github.com/kif-framework/KIF/blob/bcb58d0419cfaefe286f3df1c8618bac43b45921/Classes/KIFTestCase.m#L67) would never run.

As @glentregoning just said, we have managed to solve this issue in [KIF-Kiwi](https://github.com/indiegogo/KIF-Kiwi) by extending `KWSpec` to replicate the same behaviour on `KIFTestCase`, running `KIFEnableAccessibility()` on set up (see [KIF-Kiwi/#11](https://github.com/indiegogo/KIF-Kiwi/pull/11)).

Hope this gives enough insight to anyone in a similar situation. Thanks everyone for your help and guidance!

**Comment by mikelupo** [comment]
Thank you for sharing your insight.

**Comment by Kurry** [comment]
Thanks for the feedback @ecaselles it happened to me as well. What I did was inherit from `FBSnapshotTestCase` just so I could easily take snapshots during my KIF testcases. I was having issues when running tests using fastlane because the tests would fail because accessibility was not working. Adding `KIFEnableAccessibility()` fixed my issue!
